extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java

```diff
@@ -176,6 +166,37 @@ void discoverDefaultSerdeConfig(DefaultSerdeDiscoveryState discovery,
         }
     }
 
+    private void processIncomingType(DefaultSerdeDiscoveryState discovery,
+            BuildProducer<RunTimeConfigurationDefaultBuildItem> config, Type incomingType, String channelName) {
+        extractKeyValueType(incomingType, () -> {
+            Boolean existingBatchConfig = discovery.isBatchIncoming(channelName);
+            if (existingBatchConfig == null) {
+                // No pre-existing batch config, produce batch=true, batch type allowed
+                produceRuntimeConfigurationDefaultBuildItem(discovery, config,
+                        "mp.messaging.incoming." + channelName + ".batch", "true");
```

Comment:
This relies on an assumption that `extractKeyValueType` only calls this lambda when it has already determined that the consumer expects a batch. This is pretty hard to understand -- especially considering that the `extractKeyValueType` method calls this lambda `isBatchAllowed`.

I personally think it would be easier to understand if this lambda was dropped and the `keyValueTypeAcceptor` additionally accepted a `boolean` marking if the consumer expects a batch or not. (Do we have a `TriConsumer` somewhere? :-) ) The `(key, value) -> { ... }` lambda below would simply act on the `boolean` by calling `produceRuntimeConfigurationDefaultBuildItem(discovery, config, "mp.messaging.incoming." + channelName + ".batch", "true");`. No need to check if `.batch` is pre-configured or not -- we're just setting _default_ configuration.

Reply:
Indeed, I think I overcomplicated things to handle the List case. Would you think that it is useful to log a warning when the batch type is detected but a `batch=false` config exists?

Reply:
I don't think that's totally necessary, as noone will notice that warning, but it won't hurt -- if it doesn't complicate the logic too much (which I think it may).

Reply:
Ok I went for the simplified logic.